docs/hugo/content/design/ADR-2024-07-Dynamic-Export-To-Secret-Or-ConfigMap.md

```diff
@@ -0,0 +1,428 @@
+
+Since `ExportKubernetesResources` is already used for a variety of purposes, rather than change its signature I propose
+we add a new sister interface to it called `SecretGetter` for resources that have Secret APIs to call.
+This interface must be implemented manually alongside the manual implementation of the
+`KubernetesExporter` resource.
+
+The existing `KuberentesExporter` interfaces implementation should be updated to share code with the
+`SecretGetter` implementation to reduce code duplication.
+
+The signature of this interface is:
+```go
+type GetSecretsParameters struct {
+	Obj MetaObject
+	ARMClient *genericarmclient.GenericClient
+	Log logr.Logger
+
+	// Hints is the set of secrets used in the expression(s), which can be used to filter/define what 
+	// if any secrets should be retrieved
+	Hints set.Set[string]
+}
+
+type SecretGetter interface {
+	// GetSecrets retrieves the set of secrets defined by params.
+	// The resulting map is keyed by the secret key (== property name of operatorSpec.secrets.<key>)
+	// which is the same as what is referenced by the user when authoring their expressions and passed in
+	// via params.Hints
+	GetSecrets(
+		ctx context.Context,
+		params *GetSecretsParameters,
+	) (map[string]string, error)
+}
+```
+
+The generic reconciler will check for this extension like it does for others before invoking the new generic code
+to handle the `operatorSpec.formattedSecrets` field.
+
+## Decision
+
+### Aspect 1: Template language
+
+We will use [Option 2: CEL](#option-2-google-common-expression-language-cel) as it gets us the most power for the lowest 
+amount of effort and is fully customizable and already hardened for use in multitenant environments.
+
+### Aspect 2: Location
+
+We will use [Option 2: operatorSpec.formattedConfigMaps and operatorSpec.formattedSecrets](#option-2-two-new-operatorspec-properties-operatorspecformattedsecrets-and-operatorspecformattedconfigmaps).
+It avoids the problems with [Option 3](#option-3-a-new-crd-exporter) and avoids the 
+(admittedly minimal) collision risk with [Option 1](#option-1-part-of-operatorspecsecrets-and-operatorspecconfigmap)
+
+### Aspect 3: Secret access
+
+Only one option was presented and we're using it.
+
+### FAQ
+
+Q: How does this new way avoid calling the secrets APIs if the user didn't ask for any secrets to be exported? 
+Keep in mind that some resources such as ManagedCluster (at v2/api/containerservice/customizations/managed_cluster_extensions.go)
+require the ability to differentiate _which_ secrets are being requested as well because there are multiple
+secrets APIs.
+
+A: For `FormattedSecrets`, the function signature to discover secrets will have `secretHints` passed to it, 
+which will contain details about the specific secrets requested in the formatting expression. The resource-specific 
+implementation can use these hints to determine which (if any) APIs to call. Empty `secretHints` means no secrets
```

Comment:
Do we specify these APIs for each resource somewhere? How'd the operator know which API to call for each resource?

Reply:
The operator will call the existing handcrafted functions (they need some modifications) for this to work.